src/lib.rs

```diff
@@ -356,7 +352,69 @@ mod tests {
         simulation.run().unwrap()
     }
 
-    fn simulate_dkr(keys: &mut Vec<LocalKey>) {
+    fn simulate_dkr_removal(keys: &mut Vec<LocalKey>, remove_party_indexes: Vec<usize>) {
+        let mut broadcast_messages: HashMap<usize, Vec<RefreshMessage<GE>>> = HashMap::new();
+        let mut new_dks: HashMap<usize, DecryptionKey> = HashMap::new();
+        let mut refresh_messages: Vec<RefreshMessage<GE>> = Vec::new();
+        let mut party_key: HashMap<usize, LocalKey> = HashMap::new();
+
+        for key in keys.iter_mut() {
+            let (refresh_message, new_dk) = RefreshMessage::distribute(key);
+            refresh_messages.push(refresh_message.clone());
+            new_dks.insert(refresh_message.party_index, new_dk);
+            party_key.insert(refresh_message.party_index, key.clone());
+        }
+
+        for refresh_message in refresh_messages.iter() {
+            broadcast_messages.insert(refresh_message.party_index, Vec::new());
+        }
+
+        for refresh_message in refresh_messages.iter_mut() {
+            if !remove_party_indexes.contains(&refresh_message.party_index) {
+                refresh_message.remove_party_indexes = remove_party_indexes.clone();
+            } else {
+                let mut new_remove_party_indexes = remove_party_indexes.clone();
+                new_remove_party_indexes.retain(|value| *value != refresh_message.party_index);
+                refresh_message.remove_party_indexes = new_remove_party_indexes;
+            }
+
+            for (party_index, refresh_bucket) in broadcast_messages.iter_mut() {
+                if refresh_message.remove_party_indexes.contains(party_index) {
+                    continue;
+                }
+                refresh_bucket.push(refresh_message.clone());
```

Comment:
where referesh_bucket is used ?

Reply:
Each bucket represents the list of messages that a party identified by the key of the bucket will receive (it's a mutable reference to a list: Imagine something like:

1 -> [m1, m2 m3, m4, m5]
2 -> [m1, m2, m3, m4, m5]
3 -> [m1, m2, m3, m4, m5]
4 -> [m1, m2, m3, m4, m5]
5 -> [m1, m2, m3, m4, m5]

If all the parties are not removing anyone. If they want to remove 4, they will be like
1 -> [m1, m2 m3, m4, m5]
2 -> [m1, m2, m3, m4, m5]
3 -> [m1, m2, m3, m4, m5]
4 -> [m4] // it has access only to it's own refresh message
5 -> [m1, m2, m3, m4, m5]